src/backend/commands/pg_profile.c

```diff
@@ -0,0 +1,788 @@
+		if (password_reuse_max < PROFILE_UNLIMITED ||
+			password_reuse_max > PROFILE_MAX_VALID)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("invalid password reuse max: %d", password_reuse_max)));
+	}
+
+	/*
+	 * Check the pg_profile relation to be certain the profile doesn't already
+	 * exist.
+	 */
+	pg_profile_rel = table_open(ProfileRelationId, RowExclusiveLock);
+	pg_profile_dsc = RelationGetDescr(pg_profile_rel);
+
+	if (OidIsValid(get_profile_oid(stmt->profile_name, true)))
+		ereport(ERROR,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
+				 errmsg("profile \"%s\" already exists",
+	    						stmt->profile_name)));
+
+	/*
+	 * Build a tuple to insert
+	 */
+	MemSet(new_record, 0, sizeof(new_record));
+	MemSet(new_record_nulls, false, sizeof(new_record_nulls));
+
+	new_record[Anum_pg_profile_prfname - 1] =
+		DirectFunctionCall1(namein, CStringGetDatum(stmt->profile_name));
+
+	new_record[Anum_pg_profile_prffailedloginattempts - 1] =
+		Int32GetDatum(failed_login_attempts);
+	new_record[Anum_pg_profile_prfpasswordlocktime - 1] =
+		Int32GetDatum(password_lock_time);
+	new_record[Anum_pg_profile_prfpasswordlifetime - 1] =
+		Int32GetDatum(password_life_time);
+	new_record[Anum_pg_profile_prfpasswordgracetime - 1] =
+		Int32GetDatum(password_grace_time);
+	new_record[Anum_pg_profile_prfpasswordreusetime - 1] =
+		Int32GetDatum(password_reuse_time);
+	new_record[Anum_pg_profile_prfpasswordreusemax - 1] =
+		Int32GetDatum(password_reuse_max);
+	new_record[Anum_pg_profile_prfpasswordallowhashed - 1] =
+		Int32GetDatum(password_allow_hashed);
+	new_record_nulls[Anum_pg_profile_prfpasswordverifyfuncdb -1] =
+		true;
+	new_record_nulls[Anum_pg_profile_prfpasswordverifyfunc - 1] =
+		true;
+
+	/*
+	 * GetNewOidForProfile() / GetNewOrPreassignedOid() will return the
+	 * pre-assigned OID, if any, and error out if there was no pre-assigned
+	 * values in binary upgrade mode.
+	 */
+	{
+		profileid = GetNewOidForProfile(pg_profile_rel, ProfileOidIndexId,
+					    				Anum_pg_profile_oid,
+					    				stmt->profile_name);
+	}
+
+	new_record[Anum_pg_profile_oid - 1] =
+		ObjectIdGetDatum(profileid);
+
+	tuple = heap_form_tuple(pg_profile_dsc, new_record, new_record_nulls);
+
+	/*
+	 * Insert new record in the pg_profile table
+	 */
+	CatalogTupleInsert(pg_profile_rel, tuple);
+
+	InvokeObjectPostCreateHook(ProfileRelationId, profileid, 0);
+
+	heap_freetuple(tuple);
+
+	/*
+	 * Close pg_profile, but keep lock till commit;
+	 */
+	table_close(pg_profile_rel, NoLock);
+
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		Assert(stmt->type == T_CreateProfileStmt);
+		CdbDispatchUtilityStatement((Node *) stmt,
+					    				DF_CANCEL_ON_ERROR|
+					    				DF_WITH_SNAPSHOT|
+					    				DF_NEED_TWO_PHASE,
+					    				GetAssignedOidsForDispatch(),
+					    				NULL);
+
+		/* MPP-6929: metadata tracking */
+		MetaTrackAddObject(ProfileRelationId,
+				   			profileid,
+				   			GetUserId(),
+				   			"CREATE", "PROFILE");
+	}
```

Comment:
Comments `MPP-XXX` is from historic GPDB4, don't copy that in new codes.
Do we need to track metadata here?
And if  track it , should drop that object in your `drop` function, properly is `DropProfile(DropProfileStmt *stmt)
`

Reply:
OK, remains that and adds track metadata in DropProfile? Or remove all track metadata?

Reply:
Yes, should add track metadata in DropProfile.